lib/jna/clink_jna.cc

```diff
@@ -0,0 +1,114 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string>
+
+#include "clink/feature/one_hot_encoder.h"
+#include "clink/kernels/clink_kernels.h"
+#include "clink/utils/clink_utils.h"
+#include "nlohmann/json.hpp"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+double SquareAdd(double x, double y) {
+  std::unique_ptr<HostContext> host_context =
+      clink::CreateHostContext("s", tfrt::HostAllocatorType::kLeakCheckMalloc);
+  ExecutionContext exec_ctx(
+      *tfrt::RequestContextBuilder(host_context.get(), nullptr).build());
+
+  AsyncValueRef<double> x_async = MakeAvailableAsyncValueRef<double>(x);
+  Argument<double> x_arg(x_async.GetAsyncValue());
+
+  AsyncValueRef<double> y_async = MakeAvailableAsyncValueRef<double>(y);
+  Argument<double> y_arg(y_async.GetAsyncValue());
+
+  AsyncValueRef<double> result_async = clink::SquareAdd(x_arg, y_arg, exec_ctx);
+
+  exec_ctx.host()->Await(result_async.CopyRCRef());
+  return result_async.get();
+}
+
+double Square(double x) { return clink::Square(x); }
+
+// Struct representation of clink::SparseVector. It is only used for JNA to
+// transmit data between Java and C++.
+typedef struct SparseVectorJNA {
+  SparseVectorJNA(clink::SparseVector &);
+  ~SparseVectorJNA();
+
+  // Total dimensions of the sparse vector.
+  int n;
+  int *indices;
+  double *values;
+  // Length of indices and values array.
+  int length;
+} SparseVectorJNA;
+
+SparseVectorJNA::SparseVectorJNA(clink::SparseVector &sparse_vector) {
+  this->n = sparse_vector.size();
+  this->length = 0;
+  for (int i = 0; i < this->n; i++) {
+    if (sparse_vector.get(i) != 0.0) {
+      this->length++;
+    }
+  }
+
+  this->indices = new int[this->length];
+  this->values = new double[this->length];
+  int offset = 0;
+  for (int i = 0; i < this->n; i++) {
+    if (sparse_vector.get(i) != 0.0) {
+      this->indices[offset] = i;
+      this->values[offset] = sparse_vector.get(i);
+      offset++;
+    }
+  }
+}
+
+SparseVectorJNA::~SparseVectorJNA() {
+  delete[] this->indices;
+  this->indices = NULL;
+  delete[] this->values;
+  this->values = NULL;
+}
+
+SparseVectorJNA OneHotEncoderModel_transform(clink::OneHotEncoderModel *model,
+                                             int x, int columnIndex) {
+  clink::SparseVector sparse_vector = model->transform(x, columnIndex);
+  return SparseVectorJNA(sparse_vector);
+}
+
+clink::OneHotEncoderModel *
+OneHotEncoderModel_loadFromMemory(const char *paramsJson,
+                                  const char *modelDataProtobuf,
+                                  int modelDataProtobufLen) {
+  clink::OneHotEncoderModel *model = new clink::OneHotEncoderModel();
+
+  nlohmann::json j = nlohmann::json::parse(paramsJson);
+  model->setDropLast(strcmp(j["dropLast"].get<std::string>().c_str(), "true") ==
```

Comment:
nits: This line is a bit too long and complicated. It is probably more readable to move `strcmp(...) == 0` to a separate line.

Reply:
OK. I'll make the change.